feat: update profiler charts

keita-determined opened the pull request:
## Description

WEB-662

<!---
Lead with the intended commit body in this description field. For breaking
changes, please include "BREAKING CHANGE:" at the beginning of your commit
body.  At a minimum, this section should include a bracketed reference to the
Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly
into the description field.
-->



## Test Plan
Need to enable chart flag `?f_chart=on` in url
- Check if profiler charts look the same as before switching the chart component
- Check if data is correct
<!---
Describe the situations in which you've tested your change, and/or a screenshot
as appropriate.  Reviewers may ask questions about this test plan to ensure
adequate manual coverage of changes.
-->



## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for
particularly tricky bits of code that could use extra scrutiny, historical
context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->



## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
<!---
Retain the relevant line and replace 000 with ticket number.

DET-000
MLG-000
WEB-000
DESIGN-000
No Ticket
--->

mapmeld commented on webui/react/src/components/kit/LineChart.tsx:
```diff
@@ -18,14 +18,14 @@ import css from './LineChart.module.scss';
 export interface Serie {
   color?: string;
-  data: Record<string, [number, number][]>;
+  data: Partial<Record<XAxisDomain, [x: number, y: number][]>>;
```
🙏

mapmeld commented on webui/react/src/pages/TrialDetails/Profiles/Charts/SystemMetricChart.tsx:
```diff
@@ -90,7 +85,7 @@ const SystemMetricChart: React.FC<ChartProps> = ({ getOptionsForMetrics, trial }
       }
       title="System Metrics">
-      <UPlotChart data={systemMetrics.data} options={options} />
+      <LineChart series={systemMetrics.data} xAxis="Time" />
```
Can this be `XAxisDomain.Time` ?

keita-determined replied:
According to the original chart [latest-master](http://latest-master.determined.ai:8080/det/experiments/1783/profiler?agentId=i-0c5355850449c0f1b&name=disk_throughput_write), x-axis is date time, so i used `Time`

mapmeld replied:
in this case I think you want `xLabel="Time" xAxis={XAxisDomain.Time}`, maybe it would work with just `xAxis={XAxisDomain.Time}`

keita-determined replied:
ah i got what you meant. ok

mapmeld commented on webui/react/src/pages/TrialDetails/Profiles/Charts/SystemMetricChart.tsx:
```diff
@@ -90,7 +86,7 @@ const SystemMetricChart: React.FC<ChartProps> = ({ getOptionsForMetrics, trial }
       }
       title="System Metrics">
-      <UPlotChart data={systemMetrics.data} options={options} />
+      <LineChart series={systemMetrics.data} xAxis={XAxisDomain.Time} xLabel="Time" />
```
We should add a Y label, `yLabel={getUnitForMetricName( ... )}`  function is in `src/pages/TrialDetails/Profiles/utils.ts`

Also seems to be an issue with width of the chart and axis labels with these larger numbers, here's an example with disk_throughput_read

<img width="605" alt="Screen Shot 2023-01-27 at 12 57 01 PM" src="https://user-images.githubusercontent.com/643918/215172226-d0991e67-18b8-472f-8851-a794f31d7403.png">

keita-determined replied:
fixed

trentwatt reviewed (changes requested):
Looks good!

you are not using `data.timestamps` as the x values any more. it looks like `data.timestamps` has no actual impact on what the `useProfilerMetrics` returns.  would be nice to get rid of this dead code (i.e. everything involving `data.timestamps`).

Also, couple UI issues

the x-range is briefly reset when there is new data (see vid). i think this is an issue with the underlying component, so probably not to be addressed in this PR, but wanted to call it out.

https://user-images.githubusercontent.com/28383080/215179648-aca104ef-e070-40ad-a2d0-7e1cafa65781.mov

missing y axis label
before:
<img width="234" alt="image" src="https://user-images.githubusercontent.com/28383080/215180128-812e95b4-d913-41b1-bf6f-fb929f538ff0.png">
after:
<img width="234" alt="image" src="https://user-images.githubusercontent.com/28383080/215180157-85691002-3cea-4328-98d4-8e16f42770bb.png">

x axis ticks are a lot messier to read
<img width="476" alt="image" src="https://user-images.githubusercontent.com/28383080/215180188-3cc5296c-990d-409e-b0f1-26c834157d68.png">

i think this line (not in your PR) might be the reason: https://github.com/determined-ai/determined/blob/afd39f18dcb610f712d37450636ef1b5d0dac5af/webui/react/src/components/kit/LineChart.tsx#L185

trentwatt commented on webui/react/src/pages/TrialDetails/Profiles/Charts/TimingMetricChart.tsx:
```diff
@@ -17,10 +18,11 @@ export const TimingMetricChart: React.FC<ChartProps> = ({ trial, getOptionsForMe
+        series={timingMetrics.data}
+        xAxis={XAxisDomain.Time}
+        xLabel="Time"
+        yLabel={options.axes?.[1].label ?? ''}
```
since we are ignoring the rest of what `getOptions` returns, can we change that to just be a `getAxis`?

trentwatt commented on webui/react/src/pages/TrialDetails/Profiles/Charts/ThroughputMetricChart.tsx:
```diff
@@ -24,7 +25,12 @@ const ThroughputMetricChart: React.FC<ChartProps> = ({ getOptionsForMetrics, tri
-      <UPlotChart data={throughputMetrics.data} options={options} />
+      <LineChart
+        series={throughputMetrics.data}
+        xAxis={XAxisDomain.Time}
```
In order to get the right timestamps format for the x-axis tick labels, looks like we would need to be able to pass additional info here besides just the name. either that, or make is so where linechart changes the timestamps to this format when the x-range is small.

mapmeld replied:
+1 to adapting when x-range is small

keita-determined replied:
Talked to @mapmeld, separate it out to a new ticket

keita-determined commented:
> x axis ticks are a lot messier to read

i think its from `LineChart` component.  I didnt add something new here. let me check

trentwatt commented:
> > x axis ticks are a lot messier to read
> 
> i think its from `LineChart` component. I didnt add something new here. let me check

yeah, this is the code that's responsible for the x-axis ticks: 

https://github.com/determined-ai/determined/blob/afd39f18dcb610f712d37450636ef1b5d0dac5af/webui/react/src/pages/TrialDetails/Profiles/Profiler.tsx#L50-L52

But `LineChart` currently doesnt accept props for the x-axis besides just the name. so we would either need to have `LineChart` accept additional props, or have have it format timestamps this way when the `XAxisDomain = time` and the x range is on the order of minutes.

mapmeld commented:
Would it be possible for code inside LineChart component to look at min and
max x for time and smartly set the x axis config for that? Instead of doing
the same outside and sending in a prop.  I'm on my phone but I'll look at
the code soon

On Fri, Jan 27, 2023, 1:50 PM Trent Watson ***@***.***> wrote:

> x axis ticks are a lot messier to read
>
> i think its from LineChart component. I didnt add something new here. let
> me check
>
> yeah, this is the code that's responsible for the x-axis ticks:
>
>
> https://github.com/determined-ai/determined/blob/afd39f18dcb610f712d37450636ef1b5d0dac5af/webui/react/src/pages/TrialDetails/Profiles/Profiler.tsx#L50-L52
>
> But LineChart currently doesnt accept props for the x-axis besides just
> the name. so we would either need to have LineChart accept additional
> props, or have have it format timestamps this way when the XAxisDomain =
> time and the x range is on the order of minutes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/determined-ai/determined/pull/5852#issuecomment-1407003756>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE5GTUKBDGX7FMFBB6TJBTWUQRJRANCNFSM6AAAAAAUI7G2ZU>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

trentwatt commented:
> Would it be possible for code inside LineChart component to look at min and max x for time and smartly set the x axis config for that? Instead of doing the same outside and sending in a prop. I'm on my phone but I'll look at the code soon
> […](#)
> On Fri, Jan 27, 2023, 1:50 PM Trent Watson ***@***.***> wrote: x axis ticks are a lot messier to read i think its from LineChart component. I didnt add something new here. let me check yeah, this is the code that's responsible for the x-axis ticks: https://github.com/determined-ai/determined/blob/afd39f18dcb610f712d37450636ef1b5d0dac5af/webui/react/src/pages/TrialDetails/Profiles/Profiler.tsx#L50-L52 But LineChart currently doesnt accept props for the x-axis besides just the name. so we would either need to have LineChart accept additional props, or have have it format timestamps this way when the XAxisDomain = time and the x range is on the order of minutes. — Reply to this email directly, view it on GitHub <[#5852 (comment)](https://github.com/determined-ai/determined/pull/5852#issuecomment-1407003756)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAE5GTUKBDGX7FMFBB6TJBTWUQRJRANCNFSM6AAAAAAUI7G2ZU> . You are receiving this because your review was requested.Message ID: ***@***.***>

@mapmeld Yeah was just chatting with @keita-determined and that was the solution we were leaning towards. i noticed a similar thing when changing x-axis to time on the learning curve (single trial), so i think a solve-it-once approach is ideal.

keita-determined commented:
<img width="1257" alt="image" src="https://user-images.githubusercontent.com/109113616/215234992-2332c314-6fe5-4f13-85b0-e9141ddd577b.png">

this could be the one. i'll check it out next week

keita-determined commented:
updated axis format

<img width="1453" alt="image" src="https://user-images.githubusercontent.com/109113616/215415667-112e24fc-c69b-4242-882a-80ad914da2f4.png">

--- 

> Also seems to be an issue with width of the chart and axis labels with these larger numbers, here's an example with disk_throughput_read

fixed

<img width="84" alt="image" src="https://user-images.githubusercontent.com/109113616/215416608-d03b94e2-8559-43b1-b6a4-e7d62eeae170.png">

keita-determined commented on webui/react/src/components/kit/LineChart.tsx:
```diff
@@ -70,6 +72,8 @@ export const LineChart: React.FC<Props> = ({
   xAxis = XAxisDomain.Batches,
   xLabel,
   yLabel,
+  xTickValues = undefined,
+  yTickValues = undefined,
```
these allows us to format axis tick values

mapmeld replied:
wouldn't undefined already be the default value for these, without the `= undefined`?

mapmeld replied:
Does it make sense to send `xTickValues` and `yTickValues` from the SystemMetricChart, if we probably want this for other charts with a time x-axis?

keita-determined commented on webui/react/src/pages/TrialDetails/Profiles/Charts/SystemMetricChart.tsx:
```diff
@@ -53,10 +55,7 @@ const SystemMetricChart: React.FC<ChartProps> = ({ getOptionsForMetrics, trial }
-    () => getOptionsForMetrics(settings.name ?? '', systemMetrics.names),
-    [getOptionsForMetrics, settings.name, systemMetrics.names],
-  );
+  const yLabel = useMemo(() => getUnitForMetricName(settings.name ?? ''), [settings.name]);
```
make sense.

mapmeld reviewed (approved):
LGTM, with x-axis to be updated in the future ticket

mapmeld reviewed (approved):
✅ LGTM with feature flag now on
